File: WikiCodeParser/Tags/SpoilerTag.cs
using System.Collections.Generic;
using WikiCodeParser.Nodes;

namespace WikiCodeParser.Tags
{
    public class SpoilerTag : Tag
    {
        public SpoilerTag()
        {
            Token = "spoiler";
            Element = "span";
            ElementClass = "spoiler";
            MainOption = "text";
            Options = new[] {"text"};
            AllOptionsInMain = true;
        }

        public override INode FormatResult(Parser parser, ParseData data, State state, string scope, Dictionary<string, string> options, string text)
        {
            var visibleText = "Spoiler";
            if (options.ContainsKey("text") && options["text"].Length > 0) visibleText = options["text"];

            var before = "<" + Element;
            if (ElementClass != null) before += " class=\"" + ElementClass + '"';
            before += $" title=\"{visibleText}\">";
            var after = "</" + Element + '>';
            return new HtmlNode(before, new SpoilerNode(visibleText, parser.ParseTags(data, text, scope, IsBlock ? "block" : "inline")), after);
        }

        private class SpoilerNode : INode
        {
            private string VisibleText { get; }
            private INode SpoilerContent { get; }

            public SpoilerNode(string visibleText, INode spoilerContent)
            {
                VisibleText = visibleText;
                SpoilerContent = spoilerContent;
            }

            public string ToHtml()
            {
                return SpoilerContent.ToHtml();
            }

            public string ToPlainText()
            {
                return $"[{VisibleText}](spoiler text)";
            }

            public IEnumerable<INode> GetChildren()
            {
                yield return SpoilerContent;
            }
        }
    }
}

File: php/src/Tags/SpoilerTag.php
<?php

namespace LogicAndTrick\WikiCodeParser\Tags;

use LogicAndTrick\WikiCodeParser\Nodes\HtmlNode;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;
use LogicAndTrick\WikiCodeParser\State;

/** @noinspection PhpMultipleClassesDeclarationsInOneFile */
class SpoilerNode implements INode
{
    private string $visibleText;
    private INode $spoilerContent;

    public function __construct(string $visibleText, INode $spoilerContent)
    {
        $this->visibleText = $visibleText;
        $this->spoilerContent = $spoilerContent;
    }

    function ToHtml(): string
    {
        return $this->spoilerContent->ToHtml();
    }

    function ToPlainText(): string
    {
        return "[$this->visibleText](spoiler text)";
    }

    function GetChildren(): array
    {
        return [$this->spoilerContent];
    }

    function ReplaceChild(int $i, INode $node): void
    {
        if ($i != 0) throw new \Exception('Argument out of range');
        $this->spoilerContent = $node;
    }

    function HasContent(): bool
    {
        return true;
    }
}

class SpoilerTag extends Tag
{
    public function __construct()
    {
        parent::__construct();
        $this->token = 'spoiler';
        $this->element = 'span';
        $this->elementClass = 'spoiler';
        $this->mainOption = 'text';
        $this->options = ['text'];
        $this->allOptionsInMain = true;
    }

    public function FormatResult(Parser $parser, ParseData $data, State $state, string $scope, array $options, string $text): INode|null
    {
        $visibleText = 'Spoiler';
        if (isset($options['text']) && strlen($options['text']) > 0) $visibleText = $options['text'];

        $before = "<{$this->element}";
        if ($this->elementClass != null) $before .= ' class="' . $this->elementClass . '"';
        $before .= " title=\"{$visibleText}\">";
        $after = "</{$this->element}>";
        return new HtmlNode($before, new SpoilerNode($visibleText, $parser->ParseTags($data, $text, $scope, $this->TagContext())), $after);
    }
}